File: src/js/4_BehavioralPatterns/6_Memento/undo_redo.js
class Memento
{
  constructor(balance)
  {
    this.balance = balance;
  }
}

class BankAccount
{
  constructor(balance = 0) {
    this.balance = balance;
    this.changes = [new Memento(balance)];
    this.current = 0;
  }

  deposit(amount)
  {
    this.balance += amount;
    let m = new Memento(this.balance);
    this.changes.push(m);
    this.current++;
    return m;
  }

  restore(m)
  {
    if (m)
    {
      this.balance = m.balance;
      this.changes.push(m);
      this.current = this.changes.count - 1;
    }
  }

  undo()
  {
    if (this.current > 0)
    {
      let m = this.changes[--this.current];
      this.balance = m.balance;
      return m;
    }
    return null;
  }

  redo()
  {
    if (this.current+1 < this.changes.length)
    {
      let m = this.changes[++this.current];
      this.balance = m.balance;
      return m;
    }
    return null;
  }

  toString()
  {
    return `Balance: $${this.balance}`;
  }
}

let ba = new BankAccount(100);
ba.deposit(50);
ba.deposit(25);
console.log(ba.toString());

ba.undo();
console.log(`Undo 1: ${ba.toString()}`);
ba.undo();
console.log(`Undo 2: ${ba.toString()}`);
ba.redo();
console.log(`Redo 2: ${ba.toString()}`);

File: src/csharp/4_BehavioralPatterns/6_Memento/UndoRedo.cs
using System.Collections.Generic;
using static System.Console;

namespace DotNetDesignPatternDemos.Behavioral.Memento.UndoRedo
{
  public class Memento
  {
    public int Balance { get; }

    public Memento(int balance)
    {
      Balance = balance;
    }
  }

  public class BankAccount
  {
    private int balance;
    private List<Memento> changes = new List<Memento>();
    private int current;

    public BankAccount(int balance)
    {
      this.balance = balance;
      changes.Add(new Memento(balance));
    }

    public Memento Deposit(int amount)
    {
      balance += amount;
      var m = new Memento(balance);
      changes.Add(m);
      ++current;
      return m;
    }

    public void Restore(Memento m)
    {
      if (m != null)
      {
        balance = m.Balance;
        changes.Add(m);
        current = changes.Count - 1;
      }
    }

    public Memento Undo()
    {
      if (current > 0)
      {
        var m = changes[--current];
        balance = m.Balance;
        return m;
      }
      return null;
    }

    public Memento Redo()
    {
      if (current + 1 < changes.Count)
      {
        var m = changes[++current];
        balance = m.Balance;
        return m;
      }
      return null;
    }

    public override string ToString()
    {
      return $"{nameof(balance)}: {balance}";
    }
  }

  public class Demo
  {
    static void Main(string[] args)
    {
      var ba = new BankAccount(100);
      ba.Deposit(50);
      ba.Deposit(25);
      WriteLine(ba);

      ba.Undo();
      WriteLine($"Undo 1: {ba}");
      ba.Undo();
      WriteLine($"Undo 2: {ba}");
      ba.Redo();
      WriteLine($"Redo 2: {ba}");
    }
  }
}